### PR TITLE
feat: openfga client

### PR DESCRIPTION
**Pull request by rhamzeh**
#### I- API Feature Requests

We are getting requests in the community for API endpoints such as: `BatchCheck` ([1][1],[2][2]), `ListRelations` ([3][3], [4][4], [5][5]), non-transactional writes ([6][6]).

While those endpoints are helpful, we are avoiding pre-emptively adding them to the API for various reasons (linked in the relevant issues and discussions).

Generally what we recommend for users is that they implement some of that logic on their end. However, considering the multiple users who have requested this, it is a much better DX experience if we offer that ourselves. Integrating that functionality into the API requires that if we chose to remove it, it would be a breaking change. In the SDKs however, we can:
1. Get faster and earlier feedback for the community as we iterate on input/output and functionality
2. Easier to deprecate and do breaking changes (people who need it can other pin an earlier version of the library or copy the code we already have and have removed)
3. Allow us to provide a balance between good DX and premptively supporting too many features in the API

By moving having the `OpenFgaClient`, which is mostly a wrapper around `OpenFgaApi`, we can add support for SDK specific extensions in the client, while keeping the OpenFgaApi as mostly a compatible exact wrapper around the API

#### II- Ease of SDK Generation and Updates

Ever since we added the endpoints for creating and listing stores, every SDK has required work to
a. remove store id as the first param
b. custom patches to add missing params (e.g. see [7][7]) 

This makes it extra work for us to 1. create new sdks and 2. update existing ones as the API documentation changes, and a lot harder for new contributors to make the necessary changes when contributing new SDKs (we saw this recently with users who where trying to contribute Rust and Java SDKs)

By moving having the `OpenFgaClient`, which is mostly a wrapper around `OpenFgaApi`, we can strive to keep the auto-generated API mostly matching the API.

Initially we would keep it as is to ease migration, but the next major change can drop most of our custom work around the `OpenFgaApi`, drastically easing the burden for us and new contributors.

#### III- DX Improvements

Handling Breaking Changes:
- When possible, (e.g. with the ListObjects breaking change), the SDK could possibly support both versions in a minor version bump before dropping support for the older versions in a major version bump, thus easing migration for users.

Allows fixing some of the interfaces
- In the API `TupleKey` has all fields as optional, for the Client, we can set the fields as required based on the method passed

Better validation:
- Instead of allowing the customers to construct their own strings when parsing the tuples, we can help do that for them to prevent mistakes

### **Scope of this ticket**:
This PR is not meant to be merged! It's just to get feedback and the community. If agreed upon and approved, it will be submitted to the SDK Generator for the JS SDK initially, then implemented to the rest of the SDKs.

### **Changes**:

Most of the changes will be in this file: https://github.com/openfga/js-sdk/blob/feat/add-openfga-client/client.ts

TLDR: It exposes a new class: `OpenFgaClient`, which is a wrapper around `OpenFgaApi` and adds a bit of the required functionality.

### **Issues**:
To provide a better way to limit requests and We have two ways to parallelize requests:
1. Chucked:
  * Pros: Easy, already mostly implemented. Requires no external deps.
  * Cons: But sends requests in batches.
3. Concurrent Limit:
  * Pros: A better form of parallization, where a max of requests can be running at the same time, but does not require the first chunk to be completely processed before beginning processing another request
  * Cons: Harder to implement, the dependency we attempted to use (p-limit) is reported to only work on node, not on browsers

### **PR Status**:
Still early, tests are non-existent, functionality is broken

### **Features that are planned to be supported**:

- ListRelations
- BatchCheck
- Write (non-transactional)
- Reduces boiler plate code users have to write
- and more

### **Feedback Requested**:
Don't worry about the actual code at the moment, it is meant to be a quick initial iteration. The main focus is:
1. Is this something that provides value, and is this line of thinking the correct way forward?
2. Can the function input/output be made better? Are they enough? Can they be improved?
5. Are there any downsides to going down this path?

### **Related Tickets**:
https://github.com/openfga/sdk-generator/issues/92
https://github.com/openfga/sdk-generator/issues/23
https://github.com/openfga/openfga/issues/182
https://github.com/openfga/openfga/issues/470
https://github.com/openfga/openfga/issues/470
https://github.com/openfga/openfga/issues/361
https://github.com/openfga/community/discussions/48
https://github.com/openfga/openfga/issues/258

[1]: https://github.com/openfga/openfga/issues/182
[2]: https://github.com/openfga/openfga/issues/470
[3]: https://github.com/openfga/openfga/issues/470
[4]: https://github.com/openfga/openfga/issues/361
[5]: https://github.com/openfga/community/discussions/48
[6]: https://github.com/openfga/openfga/issues/258
[7]: https://github.com/openfga/sdk-generator/blob/main/config/clients/js/patches/add-missing-first-param.patch

### CODE REVIEW

File: client.ts [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+   * Write - Create or delete relationship tuples
+   * @param {object} body
+   * @param {objects} options
+   */
+  async write(body: ClientWriteRequest, options: ClientRequestOptsWithAuthZModelId & ClientWriteRequestOpts = {}): Promise<ClientWriteResponse> {
+    const { transaction = {} } = options;
+    const { writes, deletes } = body;
+    const authorizationModelId = this.getAuthorizationModelId(options);
+
+    if (!transaction || transaction.disable) {
```
Comment (adriantam) [inline client.ts:260]:
```suggestion
    if (transaction && !transaction.disable) {
```

File: client.ts [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    }
+
+    const results: ClientWriteResponse = { writes: [], deletes: [] };
+    await chunkSequentialCall<TupleKey, void>(
+      (chunk) => this.openFgaApi.write({ writes: { tuple_keys: chunk}, authorization_model_id: authorizationModelId })
+        .then(() => { results.writes.push(...chunk.map(tuple => ({ tuple_key: tuple, status: ClientWriteStatus.SUCCESS }))); })
+        .catch(() => { results.writes.push(...chunk.map(tuple => ({ tuple_key: tuple, status: ClientWriteStatus.FAILURE }))); }),
+      writes || [],
+      transaction.maxPerChunk || DEFAULT_MAX_METHOD_PARALLEL_REQS,
+    );
```
Comment (adriantam) [inline client.ts:285]: we will need to add some form of sleep/retry in case of 429 mechanism in between the calls.  Otherwise, the caller may get rate limited in some OpenFGA implementations.

File: client.ts [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+            _request: tuple,
+          };
+          setNotEnumerableProperty(result, "$response", response);
+          return result;
+        })
+        .catch(err => ({
+          error: err,
+          _request: tuple,
+        }))
+      )), batchCheckRequest, DEFAULT_MAX_METHOD_PARALLEL_REQS).then(results => results.flat());
```
Comment (adriantam) [inline client.ts:359]: as noted in the write case, there may be a need to add configurable sleep

File: client.ts [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+   */
+  async listRelations(listRelationsRequest: ClientListRelationsRequest, options: ClientRequestOptsWithAuthZModelId = {}): Promise<ClientListRelationsResponse> {
+    const { user, object } = listRelationsRequest;
+    let { relations } = listRelationsRequest;
+    
+    // 1- Get the list of relations on that object type
+    // 2- Filter out/error out on invalid relations
+    // 2- Call check in batch
+    const authorizationModel: AuthorizationModel | undefined = this.getAuthorizationModelId(options) ?
+      (await this.readAuthorizationModel()).authorization_model : (await this.readLatestAuthorizationModel(options)).authorization_model;
```
Comment (adriantam) [inline client.ts:408]: Can you verify whether we need to pass in the authorization model id/options to this.readAuthorizationModel?

File: client.ts [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+    }
+
+    const { type: objectType } = getObjectFromString(object);
+    const availableRelations = Object.keys(authorizationModel?.type_definitions?.find(typeDef => typeDef.type === objectType)?.relations || {});
+
+    if (!relations) {
+      relations = availableRelations;
+    }
+
+    if (relations.some(relation => !availableRelations.includes(relation))) {
```
Comment (adriantam) [inline client.ts:421]: I am a bit iffy about needing to read the authorization model itself and validate whether such model is possible in the case relations are specified.

Will it make more sense to rely on the check to throw relations not defined exception?  This reduces the need for extra reading of auth model in the common (i.e., good) path

**Review by adriantam — APPROVED** [review]
